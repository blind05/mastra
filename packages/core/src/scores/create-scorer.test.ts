import { describe, it, expect, beforeEach, vi } from 'vitest';
import {
  FunctionBasedScorerBuilders,
  MixedScorerBuilders,
  PromptBasedScorerBuilders,
} from './create-scorer.test-utils';

beforeEach(() => {
  vi.mock('crypto', () => ({
    randomUUID: () => 'test-uuid',
  }));
});

const createTestData = () => ({
  inputText: 'test input',
  outputText: 'test output',
  get userInput() {
    return [{ role: 'user', content: this.inputText }];
  },
  get agentOutput() {
    return { role: 'assistant', text: this.outputText };
  },
  get scoringInput() {
    return { input: this.userInput, output: this.agentOutput };
  },
});

describe('createScorer', () => {
  let testData: ReturnType<typeof createTestData>;

  beforeEach(() => {
    testData = createTestData();
  });

  describe('Steps as functions scorer', () => {
    it('should create a scorer with minimum steps required (analyze and generateScore) as functions', async () => {
      const scorer = FunctionBasedScorerBuilders.basic;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with synchronous functions', async () => {
      const scorer = FunctionBasedScorerBuilders.basicSync;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with mixed sync/async functions', async () => {
      const scorer = FunctionBasedScorerBuilders.mixedSyncAsync;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with preprocess step', async () => {
      const scorer = FunctionBasedScorerBuilders.withPreprocess;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with reason step', async () => {
      const scorer = FunctionBasedScorerBuilders.withReason;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with all steps', async () => {
      const scorer = FunctionBasedScorerBuilders.allSteps;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Steps as prompt objects scorer', () => {
    it('should create a scorer with minimum steps required', async () => {
      const scorer = PromptBasedScorerBuilders.basic;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with preprocess step', async () => {
      const scorer = PromptBasedScorerBuilders.withPreprocess;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with reason step', async () => {
      const scorer = PromptBasedScorerBuilders.withReason;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with all steps', async () => {
      const scorer = PromptBasedScorerBuilders.withAllSteps;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });
  });

  describe('scorers with both function and prompt objects steps', () => {
    it('should create a scorer with preprocess function and analyze prompt object', async () => {
      const scorer = MixedScorerBuilders.withPreprocessFunctionAnalzyePrompt;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with preprocess prompt and analyze function', async () => {
      const scorer = MixedScorerBuilders.withPreprocessPromptAnalyzeFunction;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with reason function and analyze prompt object', async () => {
      const scorer = MixedScorerBuilders.withReasonFunctionAnalyzePrompt;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with reason prompt and analyze function', async () => {
      const scorer = MixedScorerBuilders.withReasonPromptAnalyzeFunction;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });
  });
});
