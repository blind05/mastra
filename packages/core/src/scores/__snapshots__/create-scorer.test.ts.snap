// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createScorer > Steps as functions scorer > should create a scorer with all steps 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeStepResult": {
    "inputFromAnalyze": "TEST INPUT!",
    "outputFromAnalyze": "TEST OUTPUT!",
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": {
    "reformattedInput": "TEST INPUT",
    "reformattedOutput": "TEST OUTPUT",
  },
  "reason": "the reason the score is 1 is because the input is TEST INPUT! and the output is TEST OUTPUT!",
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with minimum steps required (analyze and generateScore) as functions 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeStepResult": {
    "input": "test input",
    "output": "test output",
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": undefined,
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with mixed sync/async functions 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeStepResult": {
    "inputFromAnalyze": "TEST INPUT!",
    "outputFromAnalyze": "TEST OUTPUT!",
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": {
    "reformattedInput": "TEST INPUT",
    "reformattedOutput": "TEST OUTPUT",
  },
  "reason": "the reason the score is 1 is because the input is TEST INPUT! and the output is TEST OUTPUT!",
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with preprocess step 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeStepResult": {
    "inputFromAnalyze": "TEST INPUT!",
    "outputFromAnalyze": "TEST OUTPUT!",
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": {
    "reformattedInput": "TEST INPUT",
    "reformattedOutput": "TEST OUTPUT",
  },
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with reason step 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeStepResult": {
    "input": "test input",
    "output": "test output",
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": undefined,
  "reason": "the reason is because the input is test input and the output is test output",
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with synchronous functions 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeStepResult": {
    "input": "test input",
    "output": "test output",
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": undefined,
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > Steps as prompt objects scorer > should create a scorer with all steps 1`] = `
{
  "analyzePrompt": "Test Analyze prompt",
  "analyzeStepResult": {
    "inputLength": 10,
    "outputLength": 11,
  },
  "preprocessPrompt": "Test Preprocess prompt",
  "preprocessStepResult": {
    "reformattedInput": "TEST INPUT",
    "reformattedOutput": "TEST OUTPUT",
  },
  "reason": "TEST REASON",
  "reasonPrompt": "Test Generate Reason prompt",
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > Steps as prompt objects scorer > should create a scorer with minimum steps required 1`] = `
{
  "analyzePrompt": "Test Analyze prompt",
  "analyzeStepResult": {
    "inputLength": 10,
    "outputLength": 11,
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": undefined,
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > Steps as prompt objects scorer > should create a scorer with preprocess step 1`] = `
{
  "analyzePrompt": "Test Analyze prompt",
  "analyzeStepResult": {
    "inputLength": 10,
    "outputLength": 11,
  },
  "preprocessPrompt": "Test Preprocess prompt",
  "preprocessStepResult": {
    "reformattedInput": "TEST INPUT",
    "reformattedOutput": "TEST OUTPUT",
  },
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > Steps as prompt objects scorer > should create a scorer with reason step 1`] = `
{
  "analyzePrompt": "Test Analyze prompt",
  "analyzeStepResult": {
    "inputLength": 10,
    "outputLength": 11,
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": undefined,
  "reason": "TEST REASON",
  "reasonPrompt": "Test Generate Reason prompt",
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > scorers with both function and prompt objects steps > should create a scorer with preprocess function and analyze prompt object 1`] = `
{
  "analyzePrompt": "Test Analyze prompt",
  "analyzeStepResult": {
    "inputLength": 10,
    "outputLength": 11,
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": {
    "reformattedInput": "TEST INPUT from preprocess function!",
    "reformattedOutput": "TEST OUTPUT from preprocess function!",
  },
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > scorers with both function and prompt objects steps > should create a scorer with preprocess prompt and analyze function 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeStepResult": {
    "inputFromAnalyze": "TEST INPUT from preprocess prompt!!",
    "outputFromAnalyze": "TEST OUTPUT from preprocess prompt!!",
  },
  "preprocessPrompt": "Test Preprocess prompt",
  "preprocessStepResult": {
    "reformattedInput": "TEST INPUT from preprocess prompt!",
    "reformattedOutput": "TEST OUTPUT from preprocess prompt!",
  },
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > scorers with both function and prompt objects steps > should create a scorer with reason function and analyze prompt object 1`] = `
{
  "analyzePrompt": "Test Analyze prompt",
  "analyzeStepResult": {
    "inputLength": 10,
    "outputLength": 11,
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": undefined,
  "reason": "the reason is because the input is 10 and the output is 11 from generateReason function",
  "runId": "test-uuid",
  "score": 1,
}
`;

exports[`createScorer > scorers with both function and prompt objects steps > should create a scorer with reason prompt and analyze function 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeStepResult": {
    "inputFromAnalyze": "test input from analyze function!",
    "outputFromAnalyze": "test output from analyze function!",
  },
  "preprocessPrompt": undefined,
  "preprocessStepResult": undefined,
  "reason": "TEST REASON",
  "reasonPrompt": "Test Generate Reason prompt",
  "runId": "test-uuid",
  "score": 1,
}
`;
